### DISCUSSION TITLE
PDF size optimisation

### DISCUSSION THREAD

**Discussion by loan-mgt** [qpdf/qpdf#1186] [Q&A] [answered]
Hi,

I am trying to optimize a PDF. Here's the current command that I am using:

```bash
qpdf --compress-streams=y --recompress-flate --optimize-images --oi-min-width=20 --oi-min-height=20 build/qpdf/A.pdf build/qpdf/out.pdf
```

The PDF comes from Figma and most of the size comes from the images (exported from Figma, 3.5MB). After using the command above, I've managed to reduce the PDF size to 1.4MB.

However, tools like ILovePdf manage to optimize it to 850KB.

From my understanding, the main difference is the image compression level.

My question is: How can I increase the image compression level using qpdf?

**Comment by jberkenbilt** [comment]
You could improve your results by adding `--compression-level=9 --object-streams=generate` to your command. You could probably do better still (since you have specified `--recompress-flate`) by using an alternative zlib. Recent versions of qpdf are tested on alternative zlib implementations. For example zlib-ng produces smaller compressed output than regular zlib. You can actually use it dynamically. Take a look at `build-scripts/test-alt-zlib` from the qpdf source tree for an example of how to do that.

**Reply by loan-mgt** [reply]
```bash
qpdf  --compress-streams=y --recompress-flate --optimize-images --oi-min-area=0 --oi-min-width=0  --oi-min-height=0 --compression-level=9 --object-streams=generate build/qpdf/A.pdf build/qpdf/out.pdf                   
```

hasn't made any meaningful difference. My understanding is that `--compression-level=9` compresses the PDF in a lossless way, but what I'm looking for to gain 70% to 80% size reduction whit a lossy image compression step. 
So, there is no current implementation of this in qpdf?

**Reply by jberkenbilt** [reply, marked as answer]
Nothing in qpdf does this right now. I wrote some code that uses GraphicsMagick and ghostscript (along with qpdf) to do this for work. At that time, I updated the pdf-invert-images example in the qpdf source tree to use the same basic approach as my work code. The code is a bit complex, and I doubt I'd have time to extract it from the system that it's part of to open up the code. The basic thrust is that the software iterates through all the images and figures out which ones would benefit from being recompressed or scaled, then it replaces the stream data for those images. The code is pretty complex to avoid writing any temporary files, but it would be possible to write a more straightforward implementation that used temporary files.
